### PR TITLE
Fix issues 433 + 429

### PR DESCRIPTION
**Pull request by pmconrad**
This PR includes
* unit tests for issues #433 and #429 
* @abitmore 's fix for #433 
* a hardfork-protected fix for #429 
* a hardfork-protected replacement for @abitmore 's recent softfork
* a new plugin for creating snapshots at a specified time or block number

The unit tests are currently failing because the softfork prevents them from being executed. In order to verify proper operation you have to check out d71936e66c9439fed339d9ec02de00103898c8ea .

The idea of the snapshot plugin is to find out the consequences of a retroactive change (the fix for #429 adjusts the current_supply of BTS). The snapshots are dumped as JSON files, which can easily be diff'd. The commit with the new plugin can with some modifications be cherry-picked on the latest release in order to create a snapshot to which you can compare.

My test at block 21100006 (don't ask) produced the following differences:
* several refund and burn workers have lower "total_burned" fields (as expected)
* current_supply of BTS is higher (as expected)
* the BTS balance of the null account is 1 satoshi lower (matches the burn worker)
* an additional empty block_summary_object (off-by-one after replacing flat_index, doesn't matter)
* a lot of transaction objects (test artifact)
* almost all budget_records differ in "from_initial_reserve" (due to different BTS supply)

### CODE REVIEW

**Comment by oxarbitrage** [comment]
this looks good, giving it a try as soon as i can.

**Comment by oxarbitrage** [comment]
started by testing the snapshot plugin, very clever, it goes over all the indexes and save the content to file in json format. very good.

when the block provided is reached the witness log will output:
```
2713568ms th_a       snapshot.cpp:84               create_snapshot      ] snapshot plugin: creating snapshot
2728249ms th_a       snapshot.cpp:112              create_snapshot      ] snapshot plugin: created snapshot
```

here is the command used to make use of the plugin:
```
programs/witness_node/witness_node --data-dir data/my-blockprod --rpc-endpoint "127.0.0.1:8090" --plugins "witness account_history snapshot" --snapshot-at-block 50000 --snapshot-to /testSnap.json
```` 

also works good by time:
```
 programs/witness_node/witness_node --data-dir data/my-blockprod --rpc-endpoint "127.0.0.1:8090" --plugins "witness account_history snapshot" --snapshot-at-time "2015-10-16T00:00:00" --snapshot-to /testSnap.json
```

I assume the snapshot creation will take some big amount of time when  executing at current block numbers to dump everything, just tried it at lower levels but i see you did it at block 21100006 and worked. No significant extra RAM is needed as the saving is object by object.

i am going after the unit tests cases now.

File: tests/tests/fee_tests.cpp [hunk 1]
```diff
@@ -983,4 +984,178 @@ BOOST_AUTO_TEST_CASE( defaults_test )
...
+
+      auto fees_to_pay = db.get_global_properties().parameters.current_fees->get<asset_create_operation>();
+
+      {
+         signed_transaction tx;
+         asset_create_operation op;
+         op.issuer = alice_id;
+         op.symbol = "ALICE";
+         op.common_options.core_exchange_rate = asset( 1 ) / asset( 1, asset_id_type( 1 ) );
+         op.fee = asset( (fees_to_pay.long_symbol + fees_to_pay.price_per_kbyte) & (~1) );
```
[the remark this answers is not in the document]
Comment (pmconrad) [reply tests/tests/fee_tests.cpp]: Yes, using (fees_to_pay.long_symbol + fees_to_pay.price_per_kbyte) is a simplification - I didn't want to make the test code more complicated than it already is.

**Comment by oxarbitrage** [comment]
@pmconrad could we included https://github.com/bitshares/bitshares-core/commit/d71936e66c9439fed339d9ec02de00103898c8ea to the pull to have everything together ?

**Comment by abitmore** [comment]
@oxarbitrage it's already included.

**Comment by oxarbitrage** [comment]
true, sorry, i think i am misunderstanding him. the test cases for `issue_433_test` and `issue_433_indirect_test` are failing. i think he meant that i need to remove the commit. if so, can we remove it from the pull and have altogether here ?

sorry i am not following it correctly. 

the other test case `issue_429_test` is currently passing with the pull as it is.

**Comment by pmconrad** [comment]
Check out d71936e instead of the branch head, then execute the 433 test cases

**Comment by oxarbitrage** [comment]
ok, got it. is this "conflict" caused because the fix by abit was merged to master directly ?

please note that while doing the checkout i can test `issue_433_test` and `issue_429_test` , they pass fine but `issue_433_indirect_test` is not there anymore.

**Comment by pmconrad** [comment]
In  d71936e issue_433_test contains both the direct test (tx with create_asset op) and the indirect test (tx with proposal containing a create_asset op). In the softfork commit, I pulled the indirect test into a separate method in order to verify that the softfork blocks them both.

**Pull request merged**